### PR TITLE
feat(rbac): implement CRUD for permission policy api

### PR DESCRIPTION
**Pull request by AndrienkoAleksandr**
### What does this pull request do
Implement CRUD Backend APIs for RBAC Policy

### Referenced issue
fixes: janus-idp/backstage-showcase#361

### Demo
https://www.youtube.com/watch?v=GfY8dTQS8i4

###  How to test it
You need extract user token and set up it:

```bash
export token=${userToken}

# Get policy:

curl http://localhost:7007/api/permission/policies/user/default/andrienkoaleksandr -H "Authorization: Bearer $token" -v

http GET http://localhost:7007/api/permission/policies/user/default/andrienkoaleksandr Authorization:"Bearer $token"

# Get policies

curl http://localhost:7007/api/permission/policies -H "Authorization: Bearer $token" -v

http GET http://localhost:7007/api/permission/policies Authorization:"Bearer $token"

# Get filtered policies

curl "http://localhost:7007/api/permission/policies?etityRef=user:default/andrienkoaleksandr&policy=read&effect=allow" -H "Authorization: Bearer $token" -v

http GET "http://localhost:7007/api/permission/policies?etityRef=user:default/andrienkoaleksandr&policy=read&effect=allow" Authorization:"Bearer $token"

# Create policy
# Let's provide read catalog policy for bitbucket user

curl -X POST "http://localhost:7007/api/permission/policies" -d '{"entityReference": "user:default/logarifm", "permission": "catalog-entity", "policy": "read", "effect":"allow"}' -H "Content-Type: application/json" -H "Authorization: Bearer $token" -v

# Let's check list policies
http GET http://localhost:7007/api/permission/policies Authorization:"Bearer $token"

# Update policy
# Let's change catalog-entity permission to policy-entity permission. Bitbucket user will lost access to read catalogs,
# but he will get access to read list permission policies
curl -X PUT "http://localhost:7007/api/permission/policies/user/default/logarifm" -d '{ "oldPolicy": { "permission": "catalog-entity", "policy": "read", "effect": "allow" }, "newPolicy": { "permission": "policy-entity", "policy": "read", "effect": "allow" } }' -H "Content-Type: application/json" -H "Authorization: Bearer $token" -v

# Let's check list policies
http GET http://localhost:7007/api/permission/policies Authorization:"Bearer $token"

# Delete policy
# Let's delete "read policy-entity" permission for bitbucket user.
curl -X DELETE "http://localhost:7007/api/permission/policies/user/default/logarifm?permission=policy-entity&policy=read&effect=allow" -H "Authorization: Bearer $token" -v
```

### CODE REVIEW

File: plugins/rh-rbac-backend/src/service/policy-builder.ts [hunk 1]
```diff
@@ -68,27 +110,308 @@ export class PolicyBuilder {
...
+function getFirstQuery(
+  queryValue: string | string[] | ParsedQs | ParsedQs[] | undefined,
+): string {
+  if (!queryValue) {
+    return '';
+  }
+  if (Array.isArray(queryValue)) {
+    if (typeof queryValue[0] === 'string') {
+      // return queryValue[0]
+      return queryValue.toString();
```
Comment (gorkem) [inline plugins/rh-rbac-backend/src/service/policy-builder.ts:399]: Is this returning the correct value?

File: plugins/rh-rbac-backend/src/service/policy-builder.ts [hunk 2]
```diff
@@ -68,27 +110,308 @@ export class PolicyBuilder {
...
+function transformPolicyArray(
+  ...policies: string[][]
+): EntityReferencedPolicy[] {
+  return policies.map((p: string[]) => {
+    const [entityReference, permission, policy, effect] = p;
+    return { entityReference, permission, policy, effect };
+  });
+}
+
+function transformPolicyToArray(policy: EntityReferencedPolicy) {
```
Comment (gorkem) [inline plugins/rh-rbac-backend/src/service/policy-builder.ts:381]: should have a return value type

File: plugins/rh-rbac-backend/src/service/policy-builder.ts [hunk 3]
```diff
@@ -68,27 +110,308 @@ export class PolicyBuilder {
...
+        policies = await enforcer.getFilteredPolicy(0, ...filter);
       } else {
-        response.send({ status: 'Authorized' });
+        policies = await enforcer.getPolicy();
       }
+
+      response.json(transformPolicyArray(...policies));
+    });
+
+    router.get('/policy/:namespace/:id', async (request, response) => {
```
Comment (gorkem) [inline plugins/rh-rbac-backend/src/service/policy-builder.ts:148]: Should we move all the router.* implemenatations to its own file?

File: plugins/rh-rbac-backend/src/service/policy-builder.ts [hunk 4]
```diff
@@ -68,27 +110,308 @@ export class PolicyBuilder {
...
+      }
+
+      resp.status(200).end();
     });
 
     return router;
   }
 }
+
+function getEntityReference(req: Request): string {
```
Comment (gorkem) [inline plugins/rh-rbac-backend/src/service/policy-builder.ts:305]: Are there unit tests for these utility methods?

**Review by gorkem — CHANGES_REQUESTED** [review]
Can you split the policy-builder.ts, see the comment for details

File: plugins/rh-rbac-backend/src/service/policy-builder.ts (hunk 3 above)
Comment (AndrienkoAleksandr) [reply plugins/rh-rbac-backend/src/service/policy-builder.ts]: Done.

File: plugins/rh-rbac-backend/src/service/policy-builder.ts (hunk 1 above)
Comment (AndrienkoAleksandr) [reply plugins/rh-rbac-backend/src/service/policy-builder.ts]: Nice catch, It was correct only for single value query. I fixed it.

Comment (AndrienkoAleksandr) [reply plugins/rh-rbac-backend/src/service/policy-builder.ts]: Also I applied tests.

File: plugins/rh-rbac-backend/src/service/policy-builder.ts (hunk 4 above)
Comment (AndrienkoAleksandr) [reply plugins/rh-rbac-backend/src/service/policy-builder.ts]: I think it make sense to align our api urls with catalog entity service: https://github.com/backstage/backstage/blob/41a2601ef613ad90e1900566bbbef80646e73e5c/plugins/catalog-backend/src/service/createRouter.ts#L190
Regarding unit tests, I believe we have adequately covered the invocation of this method by utilizing various URLs in our REST API tests.

**Pull request merged**